Google.Api.Gax.Grpc/Rest/RestChannel.cs

```diff
@@ -60,9 +62,18 @@ public RestChannel(RestServiceCollection serviceCollection, string endpoint, Cha
             var restMethod = _serviceCollection.GetRestMethod(method);
 
             var cancellationTokenSource = new CancellationTokenSource();
+            if (options.Deadline.HasValue)
+            {
+                // TODO: [virost, 2021-12] Use IClock.
+                var delayInterval = options.Deadline.Value - DateTime.UtcNow;
+                if (delayInterval.TotalMilliseconds <= 0)
+                {
+                    throw new TimeoutException($"The timeout was reached when calling a method `{restMethod.FullName}`");
```

Comment:
Hmmm, if we were going to be that strict, then we should continue to inspect the deadline even after the HTTP request returned succesfully, i.e. while we parse the response, etc.
I think a deadline so small that we cannot even do prep for the part of the process that takes the significant amount of time could be considered a bug, and different from a "really small deadline" that some times will not be enough to complete the time consuming part of the process. And I think we could signal that differently.

But I'm happy to disagree on this.

Reply:
It's not a matter of being strict as such - we're only throwing something because we know the deadline has already passed before we make the RPC; this is only a check to make sure we don't pass an invalid argument to `Task.Delay`.